contracts/TrueCurrencies/TrueRewardBackedToken.sol

```diff
@@ -230,153 +229,95 @@ contract TrueRewardBackedToken is RewardTokenWithReserve {
 
-        _value = _transferWithRewards(_from, finalTo, _value);
+        depositWithReserve(receiver, depositAmount, _toRewardToken(depositAmount, opportunity()), opportunity());
+    }
 
-        // emit transfer event for from
-        emit Transfer(_from, _to, _value);
-        if (finalTo != _to) {
-            emit Transfer(_to, finalTo, _value);
-            if (hasHook) {
-                TrueCoinReceiver(finalTo).tokenFallback(_to, _value);
-            }
-        } else {
-            if (hasHook) {
-                TrueCoinReceiver(finalTo).tokenFallback(_from, _value);
-            }
+    function _transferWithRewards(
+        address _from,
+        address _to,
+        uint256 _value
+    ) internal returns (address) {
+        if (neitherSideHasTrueRewards(_from, _to) || isOpportunity(_from) || isOpportunity(_to)) {
+            require(super.balanceOf(_from) >= _value, "not enough balance");
+            return super._transferAllArgs(_from, _to, _value);
         }
+
+        require(balanceOf(_from) >= _value, "not enough balance");
+
+        uint redeemedAmount = redeemFromSender(_from, _value);
+        address finalTo = super._transferAllArgs(_from, _to, redeemedAmount);
+        depositForReceiver(finalTo, redeemedAmount);
+
+        return finalTo;
     }
 
-    /**
-     * @dev TransferFromAll helper function for TrueRewardBackedToken
-     */
-    function _transferFromAllArgs(
+    function _transferFromWithRewards(
         address _from,
         address _to,
         uint256 _value,
         address _spender
-    ) internal {
-        // 1. Both sender and receiver are disabled -> normal transfer
-        if (!trueRewardEnabled(_from) && !trueRewardEnabled(_to)) {
-            super._transferFromAllArgs(_from, _to, _value, _spender);
-            return;
+    ) internal returns (address) {
+        if (neitherSideHasTrueRewards(_from, _to) || _to == opportunity() || _from == opportunity()) {
+            require(super.balanceOf(_from) >= _value, "not enough balance");
+            return super._transferFromAllArgs(_from, _to, _value, _spender);
         }
 
-        // check balance
         require(balanceOf(_from) >= _value, "not enough balance");
 
-        (address finalTo, bool hasHook) = _requireCanTransferFrom(_spender, _from, _to);
+        uint redeemedAmount = redeemFromSender(_from, _value);
+        address finalTo = super._transferFromAllArgs(_from, _to, redeemedAmount, _spender);
+        depositForReceiver(finalTo, redeemedAmount);
 
-        // call transfer helper
-        _value = _transferWithRewards(_from, finalTo, _value);
+        return finalTo;
+    }
 
-        // sub allowance of spender
-        _subAllowance(_from, _spender, _value);
+    /**
+     * @dev Transfer helper function for TrueRewardBackedToken
+     */
+    function _transferAllArgs(address _from, address _to, uint256 _value) internal returns (address) {
+        return _transferWithRewards(_from, _to, _value);
```

Comment:
Add check here for sender/reciever both disabled & call super

Alternatively, we can eliminate _transferWithRewards() and just have all the transfer logic here, since this function is just calling another function